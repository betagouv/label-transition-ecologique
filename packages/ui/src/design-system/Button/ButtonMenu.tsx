import {cloneElement, useState} from 'react';
import {flushSync} from 'react-dom';
import {
  autoUpdate,
  FloatingNode,
  FloatingPortal,
  offset,
  Placement,
  shift,
  size,
  useClick,
  useDismiss,
  useFloating,
  useFloatingNodeId,
  useInteractions,
} from '@floating-ui/react';

import {ButtonProps} from './types';
import {Button} from './Button';
import {OpenState} from '../../utils/types';

type Props = {
  /** Le contenu du menu */
  children: React.ReactNode;
  /** Placement du menu pour l'élément floating-ui */
  menuPlacement?: Placement;
  /** Rend le composant controllable */
  openState?: OpenState;
} & ButtonProps;

/**
 * Ouverture d'un menu flottant au click sur le bouton.
 *
 * Ne pas oublier de donner une largeur fixe au menu s'il contient des élements qui peuvent être resizer comme un sélecteur avec une valeur.
 */
export const ButtonMenu = ({
  menuPlacement = 'bottom-end',
  openState,
  children,
  ...props
}: Props) => {
  const isControlled = !!openState;
  const [open, setOpen] = useState(false);

  const isOpen = isControlled ? openState.isOpen : open;

  const handleOpenChange = () => {
    if (isControlled) {
      openState.setIsOpen(!openState.isOpen);
    } else {
      setOpen(!open);
    }
  };

  const [maxHeight, setMaxHeight] = useState(null);

  const nodeId = useFloatingNodeId();

  const {refs, context, x, y, strategy} = useFloating({
    nodeId,
    open: isOpen,
    onOpenChange: handleOpenChange,
    placement: menuPlacement,
    whileElementsMounted: autoUpdate,
    middleware: [
      offset(8),
      shift(),
      size({
        apply({availableHeight}) {
          // https://floating-ui.com/docs/size
          flushSync(() => setMaxHeight(availableHeight));
        },
      }),
    ],
  });

  const {getReferenceProps, getFloatingProps} = useInteractions([
    useClick(context),
    useDismiss(context),
  ]);

  const hasScroll =
    refs.floating.current?.scrollHeight > refs.floating.current?.clientHeight;

  return (
    <>
      {cloneElement(
        <Button {...props} />,
        getReferenceProps({
          ref: refs.setReference,
        })
      )}
      <FloatingNode id={hasScroll ? nodeId : ''}>
        {isOpen && (
          <FloatingPortal>
            <div
              {...getFloatingProps({
                ref: refs.setFloating,
                style: {
                  position: strategy,
                  top: y,
                  left: x,
                  maxHeight: maxHeight - 16,
                },
                className:
                  'relative overflow-y-auto bg-white rounded-b-lg border border-grey-4 rounded-lg shadow-card',
              })}
            >
              {children}
            </div>
          </FloatingPortal>
        )}
      </FloatingNode>
    </>
  );
};
