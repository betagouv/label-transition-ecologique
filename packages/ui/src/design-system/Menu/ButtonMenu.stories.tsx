import {Meta, StoryObj} from '@storybook/react';
import {ButtonMenu} from './ButtonMenu';

const meta: Meta<typeof ButtonMenu> = {
  component: ButtonMenu,
  argTypes: {
    notificationVariant: {
      control: {type: 'select'},
    },
    variant: {
      control: {type: 'select'},
    },
    size: {
      control: {type: 'select'},
    },
    icon: {
      control: {type: 'text'},
    },
    iconPosition: {
      control: {type: 'select'},
    },
  },
  args: {
    icon: 'leaf-line',
    notificationValue: 2,
    menuAlignment: 'left',
    children: (
      <div className="!w-44 text-sm">
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Laudantium,
        dicta natus, ea voluptates enim nobis aspernatur vero debitis
        repellendus cum temporibus, praesentium neque dolore.
      </div>
    ),
  },
};

export default meta;

type Story = StoryObj<typeof ButtonMenu>;

export const Default: Story = {
  render: args => {
    return (
      <div className="mb-52">
        <ButtonMenu {...args} />
      </div>
    );
  },
};
