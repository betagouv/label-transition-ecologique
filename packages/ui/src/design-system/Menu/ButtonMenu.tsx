import {Ref, forwardRef, useEffect, useRef, useState} from 'react';
import {ButtonProps} from '../Button/types';
import {NotificationVariant} from '../Notification';
import {NotificationButton} from '../Button/NotificationButton';
import classNames from 'classnames';

/** Ouverture d'un menu flottant au click sur le bouton */

export const ButtonMenu = forwardRef(
  (
    {
      children,
      notificationValue,
      notificationVariant,
      size = 'xs',
      menuAlignment = 'right',
      menuClassName,
      onClick,
      ...props
    }: {
      /** Contenu du menu flottant */
      children: React.ReactNode;
      /** Valeur visible dans la pastille de notification */
      notificationValue?: number;
      /** Variant de la pastille de notification */
      notificationVariant?: NotificationVariant;
      /** Alignement du menu par rapport au bouton */
      menuAlignment?: 'left' | 'right';
      /** Surcouche sur la classname du menu */
      menuClassName?: string;
    } & ButtonProps,
    ref?: Ref<HTMLButtonElement | HTMLAnchorElement>
  ) => {
    const [openMenu, setOpenMenu] = useState(false);

    const containerRef: Ref<HTMLDivElement> = useRef(null);

    const handleClick = (
      evt: React.MouseEvent<HTMLButtonElement> &
        React.MouseEvent<HTMLAnchorElement>
    ) => {
      setOpenMenu(prevState => !prevState);
      onClick?.(evt);
    };

    const handleClickOutside = event => {
      if (
        containerRef?.current &&
        !containerRef?.current.contains(event.target)
      ) {
        setOpenMenu(false);
      }
    };

    useEffect(() => {
      document.addEventListener('mousedown', handleClickOutside);
      return () => {
        document.removeEventListener('mousedown', handleClickOutside);
      };
    }, []);

    return (
      <div ref={containerRef} className="mx-2 relative w-fit">
        <NotificationButton
          ref={ref}
          size={size}
          notificationValue={notificationValue || undefined}
          notificationVariant={notificationVariant}
          onClick={handleClick}
          {...props}
        />

        {openMenu && (
          <div
            className={classNames(
              'absolute z-modal p-4 bg-white border border-grey-2 rounded-md shadow-card w-fit',
              {
                'right-0': menuAlignment === 'right',
                'left-0': menuAlignment === 'left',
                'top-12': size === 'xs',
                'top-14': size === 'sm',
                'top-16': size === 'md',
                'top-20': size === 'xl',
              },
              menuClassName
            )}
          >
            {children}
          </div>
        )}
      </div>
    );
  }
);
