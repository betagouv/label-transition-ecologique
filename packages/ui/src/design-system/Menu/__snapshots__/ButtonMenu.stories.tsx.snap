// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`design-system/Menu/ButtonMenu Default smoke-test 1`] = `
<div class="mb-52">
  <div class="mx-2 relative w-fit">
    <div class="relative w-fit h-fit">
      <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-white hover:!text-white bg-primary-7 hover:!bg-primary-8 border-primary-7 hover:!border-primary-8 p-2">
        <span class="ri-leaf-line font-normal text-center h-4 w-4 text-[1rem] leading-[1rem] fill-white group-hover:fill-white">
        </span>
      </button>
      <div class="text-primary fill-primary bg-primary-3 text-xs inline-flex items-center gap-1 p-2 rounded-full border-2 border-grey-1 shadow h-6 w-6 !p-0 justify-center absolute top-0 right-0 translate-x-1/3 -translate-y-1/3">
        <span class="font-extrabold leading-none">
          2
        </span>
      </div>
    </div>
  </div>
</div>
`;
